### PR TITLE
feature: virtual container

### PR DESCRIPTION
**Pull request by grubyak**
- downgrade webpack
- adapt config
- annotate container that is virtual

### CODE REVIEW

File: src/components/Container.tsx [hunk 1]
```diff
@@ -136,34 +136,48 @@ export class Container<P extends ContainerProperties, S extends ContainerState>
...
         return <ContainerPlaceholder { ...this.placeholderProps }/>;
     }
 
-    public render() {
+    // eslint-disable-next-line max-lines-per-function
+    public render(): JSX.Element {
         let renderScript;
 
-        if (!this.props.isInEditor && this.props.aemNoDecoration){
+        if (!this.props.isInEditor || this.props.aemNoDecoration) {
```
Comment (habansal) [inline src/components/Container.tsx:155]: can you please explain why did you change this && to ||

Comment (grubyak) [reply src/components/Container.tsx:155]: given if/else determines if annotation should be included or not

the previous condition `(!this.props.isInEditor && this.props.aemNoDecoration)` means do not show annotation if "not in editor and no decoration flag is present", problem is that `this.props.aemNoDecoration` if not provided (most of the cases) - have no default value, thus it's `undefined` and then we have: `!this.props.isInEditor && undefined` -> `true/false && false` -> `false` thus no matter what's the value of `this.props.isInEditor` you will get annotation if `this.props.aemNoDecoration` is not provided

File: src/components/Container.tsx [hunk 2]
```diff
@@ -136,34 +136,48 @@ export class Container<P extends ContainerProperties, S extends ContainerState>
...
         } else {
+            const ref = React.createRef<HTMLDivElement>();
+
+            if (this.props.isInEditor) {
+                ModelManager.getData({ path: this.props.cqPath })
+                    .then((data) => {
+                        const isVirtualContainer = Object.keys(data || {}).length === 0;
+
+                        if (isVirtualContainer && ref?.current) {
+                            ref.current.setAttribute("data-virtualcontainer", "true");
```
Comment (pfauchere) [inline src/components/Container.tsx:170]: If agreed we could prefix/scope the data attribute with cq for an outcome similar to `data-cq-virtualcontainer`

Comment (pfauchere) [reply src/components/Container.tsx:170]: Also, a unit test to set a contract would be welcomed

File: src/components/Container.tsx (hunk 1 above)
Comment (pfauchere) [reply src/components/Container.tsx:155]: I don't think that this change belongs to that PR.

And I am starting to think that the condition with `&&` is actually correct but incomplete:

# The editor requires the decoration
# If the consumer wants to remove the decoration, they can do it by setting the flag to true
## They will lose the styles and attributes we set via the decoration
# Missing: When doing the same when Server-Side Rendering, then I am expecting `this.props.isInEditor` to be false all the time. And I fear that when the rendition is meant for the Editor and that the customer has set the flag `this.props.aemNoDecoration` to true, we return the wrong rendering.

File: src/components/Container.tsx [hunk 3]
```diff
@@ -136,34 +136,48 @@ export class Container<P extends ContainerProperties, S extends ContainerState>
...
+                { this.childComponents }
+              </React.Fragment>
+            );
         } else {
+            const ref = React.createRef<HTMLDivElement>();
+
+            if (this.props.isInEditor) {
+                ModelManager.getData({ path: this.props.cqPath })
+                    .then((data) => {
+                        const isVirtualContainer = Object.keys(data || {}).length === 0;
```
[the remark this answers is not in the document]
Comment (grubyak) [reply src/components/Container.tsx:167]: Isn't this exactly how it was discussed?

File: src/components/Container.tsx (hunk 2 above)
Comment (grubyak) [reply src/components/Container.tsx:170]: Unit tests will be added when rewriting the repo, as we need to fix some failing tests adapt project first.

File: src/components/Container.tsx (hunk 1 above)
Comment (grubyak) [reply src/components/Container.tsx]: reverted, but this must be revisited as it's buggy at the moment

**Comment by sharanyavinod** [comment]
Closing this PR as the approach has been updated to be handled at AEM side
cc @grubyak
